h5netcdf/core.py

```diff
@@ -790,32 +956,21 @@ def _create_child_variable(
         if self._root._h5py.__name__ == "h5py":
             kwargs.update(dict(track_order=self._parent._track_order))
 
-        # handling default fillvalues for legacyapi
-        # see https://github.com/h5netcdf/h5netcdf/issues/182
-        from .legacyapi import Dataset, _get_default_fillvalue
-
-        # enum handling
-        if fillvalue is not None and dtype is not None:
-            if isinstance(dtype, EnumType):
-                # raise if enum type is not in current group hierarchy
-                if dtype.name not in self._parent._all_enumtypes:
-                    raise TypeError(
-                        f"Unable to create variable {name!r} with EnumType {dtype!r}. EnumType not found in group {self.name!r} or it's parents."
-                    )
-                dtype = dtype.dtype
-                fillvalue = np.array(fillvalue).astype(dtype)
-        fillval = fillvalue
-        if fillvalue is None and isinstance(self._parent._root, Dataset):
-            fillval = _get_default_fillvalue(dtype)
+        # fill value handling
+        fillvalue, h5fillvalue = _check_fillvalue(self, fillvalue, dtype, np_dtype)
+
+        # use numpy dtype for h5pyd
+        if self._root._h5py.__name__ == "h5pyd":
+            h5type = np_dtype
 
         # create hdf5 variable
         self._h5group.create_dataset(
             h5name,
             shape,
-            dtype=dtype,
+            dtype=h5type,
```

Comment:
Using the h5type (for user types this would be the committed type) this will link/share the type in the dataset.